src/axom/sidre/examples/sidre_external_array.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2017-2020, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+// Axom includes
+#include "axom/core.hpp"  // for Axom types and macros
+#include "axom/sidre.hpp" // for sidre
+#include "axom/slic.hpp"  // for logging with slic
+
+// MPI includes
+#include <mpi.h>
+
+// C/C++ includes
+#include <iostream>
+
+// aliases
+namespace sidre = axom::sidre;
+namespace slic  = axom::slic;
+
+//------------------------------------------------------------------------------
+void sidre_write( MPI_Comm comm,
+                  const std::string& file,
+                  int* data,
+                  axom::IndexType numTuples,
+                  axom::IndexType numComponents )
+{
+  SLIC_ASSERT( comm != MPI_COMM_NULL );
+  SLIC_ASSERT( data != nullptr );
+  SLIC_ASSERT( !file.empty( ) );
+
+  int nranks = -1;
+  MPI_Comm_size( comm, &nranks );
+
+  sidre::DataStore ds;
+  sidre::Group* root = ds.getRoot();
+  sidre::View*  view = root->createView( "data" );
+
+  sidre::IndexType shape[2];
+  shape[ 0 ] = numTuples;
+  shape[ 1 ] = numComponents;
+
+  view->setExternalDataPtr( sidre::INT32_ID, 2, shape, data );
+
+// DEBUG
+  SLIC_INFO( "Here is the data that is begin dumped:" );
+  root->print();
+  std::cout << std::endl;
+// DEBUG
+
+
+  // STEP 2: save the array data in to a file
+  sidre::IOManager sidre_io( comm );
+  sidre_io.write( root, nranks, file, "sidre_hdf5" );
+}
+
+//------------------------------------------------------------------------------
+void sidre_read( MPI_Comm comm,
+                 const std::string& file,
+                 int*& data,
+                 axom::IndexType& numTuples,
+                 axom::IndexType& numComponents )
+{
+  SLIC_ASSERT( comm != MPI_COMM_NULL );
+  SLIC_ASSERT( data == nullptr );
+  SLIC_ASSERT( !file.empty( ) );
+
+  int nranks = -1;
+  MPI_Comm_size( comm, &nranks );
+
+  sidre::DataStore ds;
+  sidre::Group* root = ds.getRoot();
+
+  sidre::IOManager sidre_io( comm );
+  sidre_io.read( root, file );
+
+// DEBUG
+  SLIC_INFO( "Here is the data that was read back:" );
+  root->print();
```

Comment:
However, when I read the data back, I get this:
```
{
  "name": "",
  "views":
  {
    "data":
    {
      "name": "data",
      "schema": "{\"dtype\":\"int32\", \"number_of_elements\": 40, \"offset\": 0, \"stride\": 4, \"element_bytes\": 4, \"endianness\": \"little\"}",
      "value": "null",
      "state": "EXTERNAL",
      "is_applied": 0
    }
  }
}
```

The view is there and has the correct shape information, but, where did the data go?

Do I need to mark external data some how before I write them to a file?

@nselliott, @cyrush , @rhornung67  Thoughts?

Reply:
> Do I need to mark external data some how before I write them to a file?

Actually the special steps are on the reading side, not the writing side.

Specifically, there is a three-step process that you have to do to load in external data:
1. Read in the data store
2. The host application then allocates any necessary external arrays and informs sidre about their locations
3. We then load the external data into the supplied arrays.

See the example here: https://github.com/LLNL/axom/blob/develop/src/axom/sidre/tests/spio/spio_parallel.cpp

// 1. Load datastore from file:
https://github.com/LLNL/axom/blob/1827916b4741f3d7402397a4bfd98d132cd3b656/src/axom/sidre/tests/spio/spio_parallel.cpp#L327-L335

// 2a. Host code allocates storage
https://github.com/LLNL/axom/blob/1827916b4741f3d7402397a4bfd98d132cd3b656/src/axom/sidre/tests/spio/spio_parallel.cpp#L337-L342

// 2b. Host code tells sidre about these pointers
https://github.com/LLNL/axom/blob/1827916b4741f3d7402397a4bfd98d132cd3b656/src/axom/sidre/tests/spio/spio_parallel.cpp#L344-L348

// 3. Sidre loads data into these arrays
https://github.com/LLNL/axom/blob/1827916b4741f3d7402397a4bfd98d132cd3b656/src/axom/sidre/tests/spio/spio_parallel.cpp#L350

This somewhat complicated process is necessary since sidre does not own or allocate storage associated with the external views.

Reply:
Thanks for the clarification @kennyweiss 

Honestly, this is very confusing and unintuitive. I would have expected that when the data is read back is no longer `EXTERNAL`

How do things work with just `conduit`?

Reply:
I don't disagree about it being complicated, but it is not unnecessarily complicated.

It was designed to follow an expected use case:  If a host code is using ``sidre`` for checkpoints/restarts on EXTERNAL data, it wants to control the data management for these arrays.  Converting EXTERNAL data to ``sidre``-managed data would break this design.

This whole process is described in the sidre user docs: 
https://axom.readthedocs.io/en/develop/axom/sidre/docs/sphinx/parallel_io_concepts.html#iomanager-class-use

Reply:
Would it suffice if `EXTERNAL` was an attribute ? Wouldn't that be cleaner ? 

I find this way too complicated for what is necessary. I am surprised no one else has complained about this. Most likely though not a lot of folks have run into this

Reply:
the three step process is necessary if you want the host code to own the external pointers:

 - read in all sidre data + descriptions of external views
 - allow host code  to alloc pointers for the external views and attach them
 - read in all external data in one operation

The second step has to be up to the host code - the other two are each one function call.

The external case is more complex in that it gives the host code extra control over allocate.

Reply:
An attribute still doesn't tell where the application-owned data allocation is.  There would still need to be three steps:  The DataStore loads its tree, which it has to do before it knows there are any external Views at all; the application associates its allocated adrresses with the external Views; something puts the correct data values at the application-owned adresses.

Reply:
If a code doesn't care about preserving the external ownership of the pointers, the protocols that don't have "sidre" in their names load the data into DataStore-owned memory, if I recall correctly.

Reply:
So, in order to write a general reader without making any assumptions, you would need to traverse the tree, check for all any "external" views, allocate data and set external pointers and then do another I/O call to load the external data.

I find that very perplexing.

 I guess all I am saying is that I think there could be potential for improvement here.

Reply:
> If a code doesn't care about preserving the external ownership of the pointers, the protocols that don't have "sidre" in their names load the data into DataStore-owned memory, if I recall correctly.

Could there be an 'option' in the `read` function to load the external data and let sidre assume ownership? Maybe that is a good comprise to maintain both the flexibility of the 3-step process and ease of usability. Thoughts?

Reply:
@gzagaris if you are using external, that is correct. It's there to support use cases where the host code wants extra control. 

It's there in support fortran applications for LLNL where they already allocate their own data.

Again, step 2  is where the work is, and it happens in the host code. 
Codes are already describing their data to sidre, this is a small change beyond that to accomplish restart. 

If there is a pattern or approach that can simplify, then I am all for it.

Reply:
Thanks -- I'll create a Github issue to discuss this further and revisit this.

Reply:
@gzagaris we could potentially add another read function `readEverythingPlease` (not an actual name suggestion, just to frame the discussion)

Reply:
@cyrush  yeah, that would work too.

Reply:
See #198